proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,207 @@
+```
+
+#### Block contexts and label kinds
+
+So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts.
+
+```
+B^0 ::= val* [_] instr*
+B^k ::= catch_m{ tagaddr^? instr* }* B^k end | caught_m{ tagaddr val* } B^k end
+      | delegate{ labelidx } B^k end
+B^{k+1} ::= val* (label_n{instr*} B^k end) instr*
+```
+
+#### Throw Contexts
+
+Throw contexts don't skip over handlers, they are used to match a thrown exception with the innermost handler.
+
+```
+T ::= val* [_] instr* | label_n{instr*} T end | caught_m{ tagaddr val^n } T end
+   | frame_n{F} T end
+```
+
+Note that because `catch_n` instructions are not included above, there is always a unique maximal throw context.
+
+### Reduction of instructions
+
+Reduction steps for the new instructions or administrative instructions.
+
+```
+F; throw x  ↪  F; throw a  (if F.module.tagaddrs[x]=a)
+
+caught_m{a val^n} B^l[rethrow l] end
+  ↪ caught_m{a val^n} B^l[val^n (throw a)] end
+
+caught_m{a val^n} val^m end  ↪  val^m
+```
+
+An absent tag address in a `handler` clause (i.e., `a? = ε`) represents a `catch_all`.
+
+```
+F; val^n (try bt instr* (catch x instr'*)* (catch_all instr''*)? end)
+  ↪  F; label_m{} (catch_m{a instr'*}*{ε instr''*}? val^n instr* end) end
+  (if bt = [t1^n]→[t2^m] ∧ (F.module.tagaddrs[x]=a)*)
+
+catch_m{a? instr*}* val^m end ↪ val^m
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; caught_m{a val^n} (val^n)? instr* end
+  (if (a1? = ε ∨ a1? = a) ∧ S.tags(a).type = [t^n]→[])
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
+  (if a1? ≠ ε ∧ a1? ≠ a)
+
+S; F; catch_m T[val^n (throw a)] end
+  ↪  S; F; val^n (throw a)
+
+
+val^n (try bt instr* delegate l)
+  ↪ label_m{} (delegate{l} val^n instr* end) end
+  (if bt = [t^n]→[t^m])
+
+delegate{l} val^n end ↪ val^n
+
+label_m{} B^l[ delegate{l} T[val^n (throw a)] end ] end
+  ↪ val^n (throw a)
+```
+
+Note that the last reduction step above is similar to the reduction of `br`.
+
+### Typing rules for administrative instructions
+
+```
+S.tags[a].type = [t*]→[]
+--------------------------------
+S;C ⊢ throw a : [t1* t*]→[t2*]
+
+((S.tags[a].type = [t'*]→[])?
+ S;C, labels {result [t*], kind catch} ⊢ instr'* : [t'*?]→[t*])*
+S;C, labels [t*] ⊢ instr* : []→[t*]
+-----------------------------------------------------------------
+S;C, labels [t*] ⊢ catch_n{a? instr'*}* instr* end : []→[t*]
+
+S;C, labels [t*] ⊢ instr* : []→[t*]
+C.labels[l] = [t'*]
+------------------------------------------------------
+S;C, labels [t*] ⊢ delegate{l} instr* end : []→[t*]
+
+S.tags[a].type = [t'*]→[]
+(val:t')*
+S;C, labels {result [t*], kind catch} ⊢ instr* : []→[t*]
```

Comment:
I think `[t'*]` is only related to the tag type. The empty type there is because `caught` does not consume any values. I think similarly to [this case we discussed](https://github.com/WebAssembly/exception-handling/pull/143/files/9b02debf6a3bb03f64373eb0e56f6f857fc1e68d#r759718940).